test/Microsoft.AspNet.Mvc.Razor.Test/RazorPageTest.cs

```diff
@@ -687,96 +687,154 @@ public void IsSectionDefined_ThrowsIfPreviousSectionWritersIsNotRegistered()
+                p.HtmlEncoder = new HtmlEncoder();
+                p.WriteTagHelper(tagHelperExecutionContext);
             }, context);
             await page.ExecuteAsync();
 
             // Assert
-            Assert.Equal(expectedContent, writer.ToString());
-            Assert.Equal(2, writer.Buffer.BufferEntries.Count);
-            var expectedList = new List<object>();
-            expectedList.Add("Hello ");
-            expectedList.Add("World!");
-            Assert.Equal(expectedList, writer.Buffer.BufferEntries);
+            Assert.Equal(expected, writer.ToString());
         }
 
-        [Fact]
-        public async Task Write_ITextWriterCopyable_WritesContent_AsString()
+        [Theory]
+        // This is a scenario where GetChildContentAsync is called.
+        [InlineData(true, "HelloWorld!", "<p>HelloWorld!</p>")]
+        // This is a scenario where ExecuteChildContentAsync is called.
+        [InlineData(false, "HelloWorld!", "<p></p>")]
+        public async Task WriteTagHelper_WritesContentAppropriately(
+            bool childContentRetrieved, string input, string expected)
         {
             // Arrange
-            // This writer stores the data as a string.
-            var writer = new StringWriter();
+            var defaultTagHelperContent = new DefaultTagHelperContent();
+            var writer = new StringCollectionTextWriter(Encoding.UTF8);
             var context = CreateViewContext(writer);
-            var expectedContent = "Hello World!";
-            var contentToBeCopied = new DefaultTagHelperContent().SetContent("Hello ").Append("World!");
+            var tagHelperExecutionContext = new TagHelperExecutionContext(tagName: "p",
+                   selfClosing: false,
+                   items: new Dictionary<object, object>(),
+                   uniqueId: string.Empty,
+                   executeChildContentAsync: () => {
+                       defaultTagHelperContent.SetContent(input);
+                       return Task.FromResult(result: true);
+                   },
+                   startTagHelperWritingScope: () => { },
+                   endTagHelperWritingScope: () => defaultTagHelperContent);
+            tagHelperExecutionContext.Output =
+                new TagHelperOutput("p", new Dictionary<string, string>());
+            if (childContentRetrieved)
+            {
+                await tagHelperExecutionContext.GetChildContentAsync();
+            }
 
             // Act
             var page = CreatePage(p =>
             {
-                p.Write((ITextWriterCopyable)contentToBeCopied);
+                p.HtmlEncoder = new HtmlEncoder();
+                p.WriteTagHelper(tagHelperExecutionContext);
             }, context);
             await page.ExecuteAsync();
 
             // Assert
-            Assert.Equal(expectedContent, writer.ToString());
+            Assert.Equal(expected, writer.ToString());
         }
 
```

Comment:
Add test that validates `WriteTagHelperToAsync` writes to the provided `TextWriter`